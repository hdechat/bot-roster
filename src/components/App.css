.App {
  text-align: center;
}

.App-header {
  background-color: #bcc6cc;
  height: 50px;
  padding: 20px;
  color: #c12090;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

.GenerateTeamRoster {
  margin-top: 20px;
}

.Roster {
  margin: auto;
  width: 65%;
  height: 75vh;
  overflow-y: scroll;
}

.AddRobot {
  margin: 0 auto;
  width: 30%;
}

.generate__container {
  display: flex;
  border: solid;
  margin: 10px auto;
  padding: 10px;
  width: 90%;
}

.generate__error-message {
  color: red;
  border: solid red;
  background: #fdf;
  line-height: 30px;
  position: fixed;
  top: 75px;
  left: 15%;
}

.robot-data {
  display: flex;
  flex-direction: column;
  margin: 20px;
  text-align: left;
}

.header__category {
  width: 100%;
}

.nav {
  background: #fff;
  text-decoration: none;
  border: solid;
  border-radius: 3px;
  padding: 5px;
  margin: 5px;
}

select {
  border-radius: 0;
  width: 80%;
  height: 30px;
}

button {
  background: grey;
  color: #fff;
  height: 30px;
}

input {
  height: 25px;
}

table {
  border: solid;
  width: 100%;
  margin-bottom: 40px;
  padding-bottom: 5px;
}

tbody {
  height: 20px;
}

th, td {
  border: solid thin;
}

li {
  list-style-type: none
}

li:hover {
  color: blue;
  cursor: pointer;
}

@media screen and (max-width: 1050px) {
  body {
    width: 100%;
  }

  .Roster {
    width: 100%;
  }

  .AddRobot {
    width: 100%;
  }

  .generate__container {
    display: block;
  }

}
