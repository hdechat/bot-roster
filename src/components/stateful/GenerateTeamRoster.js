import React, { Component } from 'react';
import AddRobot from './AddRobot';
import Roster from '../stateless/Roster';
import * as errors from '../helpers/error-messages';

const initialState = {
  teamName: '',
  starters: [],
  subs: [],
  error: ''
};

class GenerateTeamRoster extends Component {
  constructor(props) {
    super(props);

    this.state = initialState;

    this.count = 0;
  };

  submitTeamRoster = (event) => {
    event.preventDefault();

    let rosterStatus = this.validateTeamRoster(this.state);

    if (rosterStatus === 'valid') {
      this.props.addTeamRoster({...this.state, error: ''});
      this.clearStateValues();
      this.count = 0;
    } else {
      this.setState({error: rosterStatus})
    }
  };

  validateTeamRoster({ teamName, starters, subs }) {
    if (!teamName) {
      return errors.missingName;
    } else if (starters.length !== 10) {
      return errors.badNumberOfStarters;
    } else if (subs.length !== 5) {
      return errors.badNumberOfSubs;
    } else {
      return 'valid';
    }
  }

  clearStateValues = () => {
    this.setState({ ...this.state, ...initialState });
  };

  deleteRobot = (robot) => {
    const { category, id } = robot;

    const filtered = this.state[category].filter(bot => bot.id !== id);

    this.setState({ [category]: filtered });
  };

  addRobotToTeam = (robot) => {
    const { category, firstName, lastName, speed, strength, agility } = robot;

    const totalAttrScore = this.calculateTotalScore(speed, strength, agility);
    const id = this.createId(firstName.slice(0, 2) + lastName.slice(0, 2));
    const newRobot = {...robot, id, totalAttrScore}

    let robotStatus = this.validateNewRobot(newRobot, totalAttrScore);

    if (robotStatus === 'valid') {
      this.setState({ [category]: [...this.state[category], newRobot], error: '' });
    } else {
      this.count--;
      this.setState({ error: robotStatus});
    }
  };

  validateNewRobot(robot, score) {
    const { firstName, lastName, category } = robot;
    let robotHasDupeStrength = this.checkForDuplicate('totalAttrScore', score);
    let robotHasDupeFirstName = this.checkForDuplicate('firstName', firstName);
    let robotHasDupeLastName = this.checkForDuplicate('lastName', lastName);
    let categoryFull = this.checkForCategoryFull(category);

    if (score > 100) {
      return errors.badScore;
    } else if (robotHasDupeStrength) {
      return errors.duplicateScore;
    } else if (robotHasDupeFirstName) {
      return errors.duplicateFirstName;
    } else if (robotHasDupeLastName) {
      return errors.duplicateLastName;
    } else if (categoryFull) {
      return errors.maxPlayers;
    } else {
      return 'valid';
    }
  };

  checkForCategoryFull(category) {
    if (category === 'starters') {
      return this.state.starters.length < 10 ? false : true;
    } else {
      return this.state.subs.length < 5 ? false : true;
    }
  }

  checkForDuplicate(prop, value) {
    const categories = ['starters', 'subs'];
    let duplicate;
    let count = 0;

    while(!duplicate && count < categories.length) {
      duplicate = this.state[categories[count]].find(bot => bot[prop] === value);
      count++;
    }

    return duplicate;
  }

  createId(prefix) {
    this.count++;
    let teamNumber = this.count.toString();

    prefix = teamNumber.length < 2 ? prefix += '0' : prefix;
    return prefix + teamNumber;
  }

  calculateTotalScore(speed, strength, agility) {
    return speed + strength + agility;
  }

  updateName = (robot) => {
    const updatedBots = this.state[robot.category].map(bot => {
      return bot.id === robot.id ? robot : bot;
    });

    this.setState({ [robot.category]: updatedBots })
  }

  render() {
    return (
      <div className="GenerateTeamRoster">
      <input
        required
        id="team-name"
        value={this.state.teamName} 
        onChange={(event) => this.setState({ teamName: event.target.value })}
        placeholder="Team Name" />
        <button className="generate__submit-roster" onClick={this.submitTeamRoster}>Submit Team Roster</button>
        <p className="generate__error-message">{this.state.error}</p>
        <div className="generate__container">
          <AddRobot addRobotToTeam={this.addRobotToTeam}/>
          <Roster deleteRobot={this.deleteRobot} updateName={this.updateName} team={this.state}/>
        </div>
      </div>
    );
  };
};

export default GenerateTeamRoster;
