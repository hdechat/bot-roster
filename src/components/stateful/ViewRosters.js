import React, { Component } from 'react';
import Roster from '../stateless/Roster.js';

class ViewRosters extends Component {
  constructor(props) {
    super(props);

    this.state = {team: ''}
  }
  
  render() {
    const teamsList = this.props.teams.map((team, index) => 
      <li key={team.teamName + index} onClick={()=>this.setState({ team })}><p>{team.teamName}</p></li>
    );

    return (
      <div className="ViewRosters">
        <p>Click on the team name to view the roster!</p>
        <ul>
          {teamsList}
        </ul>
        {this.state.team ? <Roster team={this.state.team} inRosters={true}/> : ''}
      </div>
    )
  }
}

export default ViewRosters;
