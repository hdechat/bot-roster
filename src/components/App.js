import React, { Component } from 'react';
import { NavLink, Route } from 'react-router-dom';
import GenerateTeamRoster from './stateful/GenerateTeamRoster';
import ViewRosters from './stateful/ViewRosters';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = { teams: [] };
  };

  addTeamRoster = (team) => {
    this.setState({ teams: [...this.state.teams, team] });
  };

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Roster Bots</h1>
        </header>
        <NavLink to='/create-roster'>Create Team Roster</NavLink>
        <NavLink to='/rosters'>View Team Rosters</NavLink>

        <Route exact path='/create-roster' render={() => {
          return <GenerateTeamRoster addTeamRoster={this.addTeamRoster} />
        }}/>
        <Route exact path='/rosters' render={() => {
          return <ViewRosters teams={this.state.teams} />
        }}/>

      </div>
    );
  };
};

export default App;
