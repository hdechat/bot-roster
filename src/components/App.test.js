import React from 'react';
import App from './App';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import GenerateTeamRoster from './stateful/GenerateTeamRoster';
import ViewRosters from './stateful/ViewRosters';

describe('App', () => {
  let wrapper = shallow(<App />);
  let mockTeam = {
    teamName: 'The Zipperheads',
    starters: [{robot: 1}],
    subs: [{robot: 2}],
    error: ''
  };

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('adds team to team array in state', () => {
    wrapper.instance().addTeamRoster(mockTeam);

    const actual = wrapper.state();
    const expected = { teams: [mockTeam]};

    expect(actual).toEqual(expected);
  });

  it('renders the GenerateTeamRoster component with the correct endpoint', () => {
    expect(wrapper.find(GenerateTeamRoster).length).toEqual(0);

    wrapper = mount(
      <MemoryRouter initialEntries={[ '/create-roster']}>
        <App />
      </MemoryRouter>
    );

    expect(wrapper.find(GenerateTeamRoster).length).toEqual(1);
  });

  it('renders the ViewRosters component with the correct endpoint', () => {
    expect(wrapper.find(ViewRosters).length).toEqual(0);

    wrapper = mount(
      <MemoryRouter initialEntries={[ '/rosters']}>
        <App />
      </MemoryRouter>
    );

    expect(wrapper.find(ViewRosters).length).toEqual(1);
  });
});
