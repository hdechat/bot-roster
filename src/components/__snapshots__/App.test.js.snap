// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
>
          <h1
                    className="App-title"
          >
                    Roster Bots
          </h1>
</header>,
        <NavLink
          activeClassName="active"
          aria-current="page"
          to="/create-roster"
>
          Create Team Roster
</NavLink>,
        <NavLink
          activeClassName="active"
          aria-current="page"
          to="/rosters"
>
          View Team Rosters
</NavLink>,
        <Route
          exact={true}
          path="/create-roster"
          render={[Function]}
/>,
        <Route
          exact={true}
          path="/rosters"
          render={[Function]}
/>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1
            className="App-title"
>
            Roster Bots
</h1>,
          "className": "App-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Roster Bots",
            "className": "App-title",
          },
          "ref": null,
          "rendered": "Roster Bots",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "activeClassName": "active",
          "aria-current": "page",
          "children": "Create Team Roster",
          "to": "/create-roster",
        },
        "ref": null,
        "rendered": "Create Team Roster",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "activeClassName": "active",
          "aria-current": "page",
          "children": "View Team Rosters",
          "to": "/rosters",
        },
        "ref": null,
        "rendered": "View Team Rosters",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/create-roster",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "exact": true,
          "path": "/rosters",
          "render": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
>
            <h1
                        className="App-title"
            >
                        Roster Bots
            </h1>
</header>,
          <NavLink
            activeClassName="active"
            aria-current="page"
            to="/create-roster"
>
            Create Team Roster
</NavLink>,
          <NavLink
            activeClassName="active"
            aria-current="page"
            to="/rosters"
>
            View Team Rosters
</NavLink>,
          <Route
            exact={true}
            path="/create-roster"
            render={[Function]}
/>,
          <Route
            exact={true}
            path="/rosters"
            render={[Function]}
/>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1
              className="App-title"
>
              Roster Bots
</h1>,
            "className": "App-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Roster Bots",
              "className": "App-title",
            },
            "ref": null,
            "rendered": "Roster Bots",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activeClassName": "active",
            "aria-current": "page",
            "children": "Create Team Roster",
            "to": "/create-roster",
          },
          "ref": null,
          "rendered": "Create Team Roster",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "activeClassName": "active",
            "aria-current": "page",
            "children": "View Team Rosters",
            "to": "/rosters",
          },
          "ref": null,
          "rendered": "View Team Rosters",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/create-roster",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/rosters",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
